include/libtorrent/packet_pool.hpp

```diff
@@ -0,0 +1,180 @@
+		time_point send_time;
+
+		// the number of bytes actually allocated in 'buf'
+		std::uint16_t allocated;
+
+		// the size of the buffer 'buf' points to
+		std::uint16_t size;
+
+		// this is the offset to the payload inside the buffer
+		// this is also used as a cursor to describe where the
+		// next payload that hasn't been consumed yet starts
+		std::uint16_t header_size;
+
+		// the number of times this packet has been sent
+		std::uint8_t num_transmissions:6;
+
+		// true if we need to send this packet again. All
+		// outstanding packets are marked as needing to be
+		// resent on timeouts
+		bool need_resend:1;
+
+		// this is set to true for packets that were
+		// sent with the DF bit set (Don't Fragment)
+		bool mtu_probe:1;
+
+	#if TORRENT_USE_ASSERTS
+		int num_fast_resend;
+	#endif
+
+		// the actual packet buffer
+		std::uint8_t buf[1];
+	};
+
+
+	struct packet_deleter
+	{
+		void operator()(packet* handle) const
+		{
+			handle->~packet();
+			std::free(handle);
+		}
+	};
+
+	using packet_ptr = std::unique_ptr<packet, packet_deleter>;
+
+	template<int ALLOCATE_SIZE>
+	struct TORRENT_EXTRA_EXPORT packet_slab
+	{
+		static const int allocate_size{ ALLOCATE_SIZE };
+
+		explicit packet_slab(std::size_t limit) : m_limit(limit) { m_storage.reserve(m_limit); }
+
+		packet_slab(const packet_slab&) = delete;
+
+		void try_push_back(packet_ptr &p)
+		{
+			if (m_storage.size() < m_limit)
+				m_storage.push_back(std::move(p));
+		}
+
+		packet_ptr alloc()
+		{
+			if (m_storage.empty())
+				return packet_ptr(static_cast<packet*>(std::malloc(sizeof(packet) + allocate_size)));
+			else
+			{
+				auto ret = std::move(m_storage.back());
+				m_storage.pop_back();
+				return ret;
+			}
+		}
+
+		void flush() { m_storage.resize(0); }
+	private:
+		const std::size_t m_limit;
+		aux::vector<packet_ptr, std::size_t> m_storage;
+	};
+
+	// single thread packet allocation packet pool
+	// can handle common cases of packet size by 3 pools
+	struct TORRENT_EXTRA_EXPORT packet_pool : private single_threaded
+	{
+		packet *acquire(int allocate)
+		{
+			TORRENT_ASSERT(is_single_thread());
+			TORRENT_ASSERT(allocate >= 0);
+			TORRENT_ASSERT(allocate <= std::numeric_limits<std::uint16_t>::max());
+
+			packet_ptr p{ alloc(allocate) };
+			new (p.get()) packet();
+			p->allocated = static_cast<std::uint16_t>(allocate);
+			return p.release();
+		}
+
+		void release(packet *p)
+		{
+			TORRENT_ASSERT(is_single_thread());
+
+			if (p == nullptr) return;
+
+			packet_ptr pp(p); // takes ownership and may auto free if slab container does not move it
+			std::size_t allocated = pp->allocated;
+
+			if (allocated <= m_syn_slab.allocate_size) { m_syn_slab.try_push_back(pp); }
+			else if (allocated <= m_mtu_floor_slab.allocate_size) { m_mtu_floor_slab.try_push_back(pp); }
+			else if (allocated <= m_mtu_ceiling_slab.allocate_size) { m_mtu_ceiling_slab.try_push_back(pp); }
+		}
+
+		//TODO: call from utp manager to flush mem
+		void flush()
```

Comment:
do you have any ideas of how to call this?
One way (which may be a bit involved) would be to have a exponential decay inside each slab. flush is called periodically and it frees some portion of the buffers in there. The slabs only contain the "head room", so it would not interfere with heavy churn of the buffers